lib/signers/v4.js

```diff
@@ -169,10 +171,13 @@ AWS.Signers.V4 = inherit(AWS.Signers.RequestSigner, {
   },
 
   hexEncodedBodyHash: function hexEncodedBodyHash() {
-    if (this.isPresigned() && this.serviceName === 's3' && !this.request.body) {
+    var request = this.request;
+    if (this.isPresigned() && this.serviceName === 's3' && !request.body) {
       return 'UNSIGNED-PAYLOAD';
-    } else if (this.request.headers['X-Amz-Content-Sha256']) {
-      return this.request.headers['X-Amz-Content-Sha256'];
+    } else if (this.operation && this.operation.authtype && this.operation.authtype.indexOf('unsigned-body') >= 0) {
+      return 'UNSIGNED-PAYLOAD';
```

Comment:
Wouldn't this case be covered by the change to event_listeners (`this.request.headers['X-Amz-Content-Sha256']` should already be `'UNSIGNED-PAYLOAD'`)?